README.md
# hypercapnia

<!-- badges: start -->
[![Project Status: WIP – Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.](https://www.repostatus.org/badges/latest/wip.svg)](https://www.repostatus.org/#wip)
[![license](http://img.shields.io/badge/license-GPL%20%28%3E=%203%29-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html)
[![R build status](https://github.com/sgibb/hypercapnia/workflows/R-CMD-check/badge.svg)](https://github.com/sgibb/hypercapnia/actions)
[![Codecov test coverage](https://codecov.io/gh/sgibb/hypercapnia/branch/main/graph/badge.svg)](https://codecov.io/gh/sgibb/hypercapnia?branch=main)
[![cran checks](https://cranchecks.info/badges/worst/hypercapnia)](https://cran.r-project.org/web/checks/check_results_hypercapnia.html)
[![metacran version](http://www.r-pkg.org/badges/version/hypercapnia)](http://cran.r-project.org/web/packages/hypercapnia/index.html)
<!-- badges: end -->

Calculate Carbon Dioxide Equivalent of Inhaled Anaesthetics.

## Install development version (not recommended)

[GitHub](https://github.com) is not directly supported by the basic
`install.packages` command. You could use the
[remotes](http://cran.r-project.org/package=remotes) package
to install the development version of `hypercapnia`.

```r
install.packages("remotes")
remotes::install_github("sgibb/hypercapnia")
```

## Contact

You are welcome to:

* submit suggestions and bug-reports at: <https://github.com/sgibb/hypercapnia/issues>
* send a pull request on: <https://github.com/sgibb/hypercapnia/pulls>
* compose an e-mail to: <mail@sebastiangibb.de>


NEWS.md
# hypercapnia

## Changes in 0.0.2

- Update GWP data to Andersen et al. 2012 and 2021.

## Changes in 0.0.1

- Initial version.


tests/testthat.R
library(testthat)
library(hypercapnia)

test_check("hypercapnia")


tests/testthat/test_conversion.R
test_that("litre_per_hour", {
    expect_equal(litre_per_hour(0.02, 0.5), 0.6)
})

test_that("gram_per_hour", {
    expect_equal(gram_per_hour(200, 0.02, 0.5), 5, tolerance = 1e-2)
})

test_that("cde", {
    expect_equal(cde_gram_per_hour(200, 0.02, 349, 2), 6964, tolerance = 1e-6)
})

test_that(".molar_volume", {
    expect_equal(.molar_volume(273.15, 101.325), 22.414, tolerance = 1e-5)
})


R/data.R
#' Fuels
#'
#' A dataset with carbon dioxid data of the some fuels.
#'
#' @format A `data.frame` with 2 rows and 2 columns:
#' \describe{
#' \item{Name}{`character`, name.}
#' \item{CO2}{`double`, gram CO2 per litre fuel.}
#' }
#'
#' @references
#' Umweltbundesamt. "Ratgeber: Sprit sparen und mobil sein" (2009).
#' \url{https://www.umweltbundesamt.de/sites/default/files/medien/publikation/long/3705.pdf}
#' accessed 2021-11-11.
"fuels"

#' Inhalational Anesthetics
#'
#' A dataset of physical properties of the some inhalational anesthetics.
#'
#' @format A `data.frame` with 3 rows and 9 columns:
#' \describe{
#' \item{Name}{`character`, name.}
#' \item{MAC}{`double`, minimum alveolar concentration.}
#' \item{MW}{`double`, molecular weight.}
#' \item{Density}{`double`, density.}
#' \item{GWP20}{`double`, global warming potential for 20 years.}
#' \item{GWP100}{`double`, global warming potential for 100 years.}
#' \item{GWP500}{`double`, global warming potential for 500 years.}
#' \item{Unit}{`numeric`, mili litre per unit/bottle.}
#' \item{Color}{`character`, color code.}
#' }
#'
#' @references
#' Andersen, Nielsen, Wallington, Karpichev, and Sander
#' "Assessing the Impact on Global Climate from General Anesthetic Gases"
#' Anesthesia & Analgesia 114.5 (2012): 1081-1085
#' \doi{10.1213/ANE.0b013e31824d6150}
#'
#' Andersen, Nielsen, and Sherman
#' "The Global Warming Potentials for Anesthetic Gas Sevoflurane Need
#' Significant Corrections"
#' Environ. Sci. Technol. 2021, 55, 15, 10189–10191
#' \doi{10.1021/acs.est.1c02573}
"ianes"


R/conversion.R
#' Litre of Volatile Anaesthetic per Hour
#'
#' @param mac `double`, minimum alveolar concentration (as ratio, e.g. for 2 %,
#' 0.02).
#' @param fgf `double`, fresh gas flow in litre per minute.
#' @return `double`, litre of volatile anaesthetic per hour.
#' @export
#' @examples
#' litre_per_hour(0.02, 0.5)
litre_per_hour <- function(mac, fgf) {
    mac * fgf * 60L
}

#' Gram of Volatile Anaesthetic per Hour
#'
#' @param mass `double`, molare mass in g/mol.
#' @param mac `double`, minimum alveolar concentration (as ratio, e.g. for 2 %,
#' 0.02).
#' @param fgf `double`, fresh gas flow in litre per minute.
#' @param temperature `double` in Kelvin.
#' @param pressure `double` in kPa.
#' @return `double`, gram of volatile anaesthetic per hour.
#' @export
#' @examples
#' gram_per_hour(200, 0.02, 0.5)
gram_per_hour <- function(mass, mac, fgf,
                         temperature = 293.15, pressure = 101.325) {
    litre_per_hour(mac, fgf) / .molar_volume(temperature, pressure) * mass
}

#' CDE
#'
#' Carbon dioxide equivalent.
#'
#' @param mass `double`, molare mass in g/mol.
#' @param mac `double`, minimum alveolar concentration (as ratio, e.g. for 2 %,
#' 0.02).
#' @param gwp `double`, global warming potential.
#' @param fgf `double`, fresh gas flow in litre per minute.
#' @param temperature `double` in Kelvin.
#' @param pressure `double` in kPa.
#' @return `double`, carbon dioxide equivalent in gram per hour.
#' @aliases cde cde_gram_per_hour
#' @export
#' @examples
#' # sevofluran, gwp20
#' cde(200, 0.02, 349, 0.5)
#'
#' # same using package provided data
#' data(ianes)
#' sevo <- subset(ianes, Name == "Sevoflurane")
#' # for 20 years
#' cde(sevo$MW, sevo$MAC, sevo$GWP20, 0.5)
#' # for 100 years
#' cde(sevo$MW, sevo$MAC, sevo$GWP100, 0.5)
cde <- function(mass, mac, gwp, fgf,
                              temperature = 293.15, pressure = 101.325) {
    gram_per_hour(
        mass, mac, fgf, temperature = temperature, pressure = pressure
    ) * gwp
}

#' @rdname cde
#' @export
cde_gram_per_hour <- cde

#' Molar Volume
#'
#' V = (nRT)/P
#'
#' @param temperature `double` in Kelvin.
#' @param pressure `double` in kPa.
#' @return `double`, molare volume at `temperature` and `pressure`.
#' @noRd
#' @examples
#' # at STP
#' .molar_volume(273.15, 101.325)
#' # at NTP
#' .molar_volume(293.15, 101.325)
.molar_volume <- function(temperature, pressure) {
    8.31446261815324 * temperature / pressure
}
